modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/planner/PlannerTimeoutTest.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.ignite.internal.processors.query.calcite.prepare.IgnitePlanner;
+import org.apache.ignite.internal.processors.query.calcite.prepare.PlanningContext;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteConvention;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteRel;
+import org.apache.ignite.internal.processors.query.calcite.schema.CacheIndexImpl;
+import org.apache.ignite.internal.processors.query.calcite.schema.IgniteSchema;
+import org.apache.ignite.internal.processors.query.calcite.trait.IgniteDistribution;
+import org.apache.ignite.internal.processors.query.calcite.trait.IgniteDistributions;
+import org.apache.ignite.internal.processors.query.calcite.trait.TraitUtils;
+import org.apache.ignite.internal.processors.query.calcite.type.IgniteTypeFactory;
+import org.apache.ignite.internal.processors.query.calcite.type.IgniteTypeSystem;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.junit.Test;
+
+import static org.apache.calcite.tools.Frameworks.createRootSchema;
+import static org.apache.calcite.tools.Frameworks.newConfigBuilder;
+import static org.apache.ignite.internal.processors.query.calcite.CalciteQueryProcessor.FRAMEWORK_CONFIG;
+
+/**
+ * Test planner timeout.
+ */
+public class PlannerTimeoutTest extends AbstractPlannerTest {
+    /** */
+    private static final long PLANNER_TIMEOUT = 1_000;
+
+    /** */
+    private static final IgniteTypeFactory typeFactory = new IgniteTypeFactory(IgniteTypeSystem.INSTANCE);
+
+    /** */
+    @Test
+    public void testLongPlanningTimeout() throws Exception {
+        SchemaPlus schema = createSchema("PUBLIC",
+            Pair.of("T1", createTestTable(Pair.of("A", Integer.class), Pair.of("B", Integer.class))),
+            Pair.of("T2", createTestTable(Pair.of("A", Integer.class), Pair.of("C", Integer.class)))
+        );
+
+        String sql = "SELECT * FROM T1 JOIN T2 ON T1.A = T2.A";
+
+        PlanningContext ctx = PlanningContext.builder()
+            .parentContext(BaseQueryContext.builder()
+                .frameworkConfig(newConfigBuilder(FRAMEWORK_CONFIG)
+                    .defaultSchema(schema)
+                    .costFactory(new IgniteCostFactory(1, 100, 1, 1))
+                    .build())
+                .logger(log)
+                .build()
+            )
+            .plannerTimeout(PLANNER_TIMEOUT)
+            .query(sql)
+            .build();
+
+        AtomicReference<IgniteRel> plan = new AtomicReference<>();
+        AtomicReference<RelOptPlanner.CannotPlanException> plannerError = new AtomicReference<>();
+
+        GridTestUtils.assertTimeout(3 * PLANNER_TIMEOUT, TimeUnit.MILLISECONDS, () -> {
+            try (IgnitePlanner planner = ctx.planner()) {
+                plan.set(physicalPlan(planner, ctx.query()));
+
+                VolcanoPlanner volcanoPlanner = GridTestUtils.getFieldValue(planner, "planner");
```

Comment:
`ctx.cluster().planner()`?

Reply:
This is private field, unfortunatelly

Reply:
But you can access this private field via `ctx.cluster().planner()`

Reply:
No, it is wrong